spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/RefreshableConfigServerIntegrationTests.java

```diff
@@ -0,0 +1,133 @@
+import org.springframework.cloud.config.server.resource.ResourceRepository;
+import org.springframework.cloud.config.server.test.ConfigServerTestUtils;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.core.io.ByteArrayResource;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.test.annotation.DirtiesContext;
+import org.springframework.test.context.ActiveProfiles;
+import org.springframework.test.context.junit4.SpringRunner;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.ArgumentMatchers.isA;
+import static org.mockito.ArgumentMatchers.nullable;
+import static org.mockito.BDDMockito.given;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest(classes = TestConfiguration.class, properties = {
+		"spring.cloud.config.enabled=true",
+		"management.endpoints.web.exposure.include=env, refresh" }, webEnvironment = WebEnvironment.RANDOM_PORT)
+@ActiveProfiles("test")
+@DirtiesContext
+public class RefreshableConfigServerIntegrationTests {
+
+	private static String localRepo = null;
+
+	@LocalServerPort
+	private int port;
+
+	@BeforeClass
+	public static void init() throws IOException {
+		// mock Git configuration to make tests independent of local Git configuration
+		SystemReader.setInstance(new MockSystemReader());
+
+		localRepo = ConfigServerTestUtils.prepareLocalRepo();
+	}
+
+	@AfterClass
+	public static void after() throws IOException {
+		ConfigServerTestUtils.deleteLocalRepo(localRepo);
+	}
+
+	@Test
+	public void refreshOverrides() throws Exception {
+		Environment environment = new TestRestTemplate().getForObject(
+				"http://localhost:" + this.port + "/foo/development/", Environment.class);
+		assertThat(environment.getPropertySources()).isEmpty();
+
+		String actuatorEndpoint = "http://localhost:" + this.port + "/actuator";
+		HttpHeaders headers = new HttpHeaders();
+		headers.set("Content-Type", "application/json");
+		HttpEntity<String> request = new HttpEntity<>(
+				"{\"name\": \"spring.cloud.config.server.overrides.foo\", \"value\": \"bar\"}",
+				headers);
+		ResponseEntity<Void> response = new TestRestTemplate()
+				.postForEntity(actuatorEndpoint + "/env", request, Void.class);
+		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
+
+		response = new TestRestTemplate().postForEntity(actuatorEndpoint + "/refresh",
+				null, Void.class);
+		assertThat(response.getStatusCode()).isEqualTo(HttpStatus.OK);
+
+		environment = new TestRestTemplate().getForObject(
+				"http://localhost:" + this.port + "/foo/development/", Environment.class);
+		assertThat(environment.getPropertySources()).isNotEmpty();
+		assertThat(environment.getPropertySources().get(0).getSource().get("foo"))
```

Comment:
What is the value of foo prior to the refresh?

Reply:
There weren't any properties at first. See [L85](https://github.com/spring-cloud/spring-cloud-config/pull/1433/files/cdb127e941056fae62b0577a32b4bea188621fb6#diff-b144d4082f116fc22971058c17bd001cR85)
It's only after setting the `overrides` and refreshing the context that the foo application will have this `foo: bar` property.

Reply:
And because it's using `spring.cloud.config.server.overrides.foo` and not just `foo` we know this works then?

Reply:
Yes. We're emulating an application `foo` which is running with the `development` profile and is asking for its properties using the REST endpoint. We're also calling the `/env` & `/refresh` actuator endpoints to change the `spring.cloud.config.server.overrides.foo` property. 
Since we see that we only get the "overridden" `foo` property after the context refresh we are sure that the properties have been set and the `EnvironmentController` bean has successfully been recreated with the new overrides.

Reply:
Awesome. Could you just copy that explanation into the test?

Reply:
Resolved with [cb85fc3](https://github.com/spring-cloud/spring-cloud-config/pull/1433/commits/cb85fc3154a80312e65d3709c9c41e94c12833dd).